clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java

```diff
@@ -385,45 +389,47 @@ public boolean handleResponse(FetchResponse<?> response) {
             } else {
                 nextMetadata = nextMetadata.nextCloseExisting();
             }
-            return false;
+            success = false;
         } else if (nextMetadata.isFull()) {
             String problem = verifyFullFetchResponsePartitions(response);
             if (problem != null) {
                 log.info("Node {} sent an invalid full fetch response with {}", node, problem);
                 nextMetadata = FetchMetadata.INITIAL;
-                return false;
+                success = false;
             } else if (response.sessionId() == INVALID_SESSION_ID) {
                 log.debug("Node {} sent a full fetch response{}",
                     node, responseDataToLogString(response));
                 nextMetadata = FetchMetadata.INITIAL;
-                return true;
+                success = true;
             } else {
                 // The server created a new incremental fetch session.
                 log.debug("Node {} sent a full fetch response that created a new incremental " +
                     "fetch session {}{}", node, response.sessionId(), responseDataToLogString(response));
                 nextMetadata = FetchMetadata.newIncremental(response.sessionId());
-                return true;
+                success = true;
             }
         } else {
             String problem = verifyIncrementalFetchResponsePartitions(response);
             if (problem != null) {
                 log.info("Node {} sent an invalid incremental fetch response with {}", node, problem);
                 nextMetadata = nextMetadata.nextCloseExisting();
-                return false;
+                success = false;
             } else if (response.sessionId() == INVALID_SESSION_ID) {
                 // The incremental fetch session was closed by the server.
                 log.debug("Node {} sent an incremental fetch response closing session {}{}",
                     node, nextMetadata.sessionId(), responseDataToLogString(response));
                 nextMetadata = FetchMetadata.INITIAL;
-                return true;
+                success = true;
             } else {
                 // The incremental fetch session was continued by the server.
                 log.debug("Node {} sent an incremental fetch response for session {}{}",
                     node, response.sessionId(), responseDataToLogString(response));
                 nextMetadata = nextMetadata.nextIncremental();
-                return true;
+                success = true;
             }
         }
+        fetchInProgress = false;
```

Comment:
Should this be done inside a `finally` block? Are we sure that this always executes after a call to `build`?

Reply:
I am not very familiar with this code but this solution seem very specific to fetch requests. Do we have a similar data race in other RPCs but we don't see it because of the limited frequency in those calls?

Reply:
agree that the proper patterns should be evaluated and applied. Request data models should be immutable once constructed, and all mutable state should be thread-safe if necessary using proper encapsulation.

Reply:
This code has been removed.